csharp/lib/src/lib.rs

```diff
@@ -79,38 +80,46 @@ pub extern "C" fn create_client(
 ) -> *const c_void {
     match create_client_internal(host, port, use_tls, success_callback, failure_callback) {
         Err(_) => std::ptr::null(), // TODO - log errors
-        Ok(client) => Box::into_raw(Box::new(client)) as *const c_void,
+        Ok(client) => Arc::into_raw(Arc::new(client)) as *const c_void,
     }
 }
 
 #[no_mangle]
 pub extern "C" fn close_client(client_ptr: *const c_void) {
-    let client_ptr = unsafe { Box::from_raw(client_ptr as *mut Client) };
-    let _runtime_handle = client_ptr.runtime.enter();
-    drop(client_ptr);
+    let count = Arc::strong_count(&unsafe { Arc::from_raw(client_ptr as *mut Client) });
+    assert!(count == 1, "Client is still in use.");
 }
 
 /// Expects that key and value will be kept valid until the callback is called.
+///
+/// # Safety
+///
+/// This function should only be called once, and should complete
+/// and call one of the response callbacks before `close_client` is called. After
+/// that the `client_ptr` is not in a valid state.
```

Comment:
Should we mention that the `client_ptr` must come from `create_client`? The [safety docs](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.increment_strong_count) for `Arc::increment_strong_count` say we must obtain the pointer from `Arc::into_raw` but it's not immediately obvious from looking at this function where that `Arc::into_raw` happens.

Also applies to the Golang file.

Reply:
done